source/common/buffer/buffer_impl.h

```diff
@@ -297,13 +311,38 @@ class Slice {
    * Call all drain trackers associated with the slice, then clear
    * the drain tracker list.
    */
-  void callAndClearDrainTrackers() {
+  void callAndClearDrainTrackersAndCharges() {
     for (const auto& drain_tracker : drain_trackers_) {
       drain_tracker();
     }
     drain_trackers_.clear();
+
+    for (auto& account : charges_) {
+      if (account) {
+        account->credit(capacity_);
+      }
+    }
+    charges_.clear();
+  }
+
+  /**
+   * Charges the provided account for the resources if we're not already
+   * charging for this slice, the account still exists, and the slice owns
+   * backing memory.
+   */
+  void chargeAccountIfValid(std::shared_ptr<Account> account) {
```

Comment:
Incrementing the account refcount is not needed in all branches of this method.  I recommend that you change the input argument type to "const std::shared_ptr<Account>&" so it is possible to avoid the increment if this call returns early.